datafusion-examples/examples/sql_analysis.rs

```diff
@@ -0,0 +1,152 @@
+    plan.apply(|node| {
+        // if we encouter a join we update the total
+        if matches!(node, LogicalPlan::Join(_) | LogicalPlan::CrossJoin(_)) {
+            total += 1;
+        }
+        Ok(TreeNodeRecursion::Continue)
+    })
+    .unwrap();
+
+    total
+}
+
+/// Counts the total amount of joins and collects every join tree with it's join count
+/// The list of groups sums to the total count
+fn count_trees(plan: &LogicalPlan) -> (usize, Vec<usize>) {
+    // this works the same way as `total_count`, but now when we encounter a Join
+    // we try to collect it's entire tree
+    let mut to_visit = vec![plan];
+    let mut total = 0;
+    let mut groups = vec![];
+
+    while let Some(node) = to_visit.pop() {
+        // if we encouter a join, we know were at the root of the tree
+        // count this tree and recurse on it's inputs
+        if matches!(node, LogicalPlan::Join(_) | LogicalPlan::CrossJoin(_)) {
+            let (group_count, inputs) = count_tree(node);
+            total += group_count;
+            groups.push(group_count);
+            to_visit.extend(inputs);
+        } else {
+            to_visit.extend(node.inputs());
+        }
+    }
+
+    (total, groups)
+}
+
+// count the entire join tree and return it's inputs using TreeNode API
+fn count_tree(join: &LogicalPlan) -> (usize, Vec<&LogicalPlan>) {
+    let mut inputs = Vec::new();
+    let mut total = 0;
+
+    join.apply(|node| {
+        // Some extra knowledge:
+        // optimized plans have their projections pushed down as far as possible, which sometimes results in a projection going in between 2 subsequent joins
+        // giving the illusion these joins are not "related", when in fact they are.
+        // just continue the recursion in this case
+        if let LogicalPlan::Projection(_) = node {
+            return Ok(TreeNodeRecursion::Continue);
+        }
+
+        // any join we count
+        if matches!(node, LogicalPlan::Join(_) | LogicalPlan::CrossJoin(_)) {
+            total += 1;
+            Ok(TreeNodeRecursion::Continue)
+        } else {
+            inputs.push(node);
+            // skip children of input node
+            Ok(TreeNodeRecursion::Jump)
+        }
+    })
+    .unwrap();
+
+    (total, inputs)
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    // To show how we can count the joins in a sql query we'll be using query 88 from the
+    // TPC-DS benchmark. It has a lot of joins, cross-joins and multiple join-trees, perfect for our example
+
+    let config = SessionConfig::default();
+    let ctx = SessionContext::new_with_config(config);
+
+    // register the tables of the TPC-DS query
+    let tables = tpcds_schemas();
+    for table in tables {
+        ctx.register_table(
+            table.name,
+            Arc::new(MemTable::try_new(Arc::new(table.schema.clone()), vec![])?),
+        )?;
+    }
+
+    let query_88 = "datafusion/core/tests/tpc-ds/88.sql";
```

Comment:
This is very cool -- thank you

This example might actually be easier to understand if we inlined these queries below -- I can do it if you like -- so the example stands on its own

https://github.com/apache/datafusion/blob/main/datafusion/core/tests/tpc-ds/88.sql